src/Quartz/Simpl/TaskSchedulingThreadPool.cs

```diff
@@ -0,0 +1,284 @@
+            if (Scheduler == null)
+            {
+                Scheduler = GetDefaultScheduler();
+            }
+
+            // Initialize the concurrency semaphore with the proper initial count
+            concurrencySemaphore = new SemaphoreSlim(MaxConcurency);
+            isInitialized = true;
+
+            log.Debug($"TaskSchedulingThreadPool configured with max concurrency of {MaxConcurency} and TaskScheduler {Scheduler.GetType().Name}.");
+        }
+
+        /// <summary>
+        /// Determines the number of threads that are currently available in
+        /// the pool; blocks until at least one is avaialble
+        /// </summary>
+        /// <returns>The number of currently available threads</returns>
+        public int BlockForAvailableThreads()
+        {
+            if (isInitialized && !shutdownCancellation.IsCancellationRequested)
+            {
+                try
+                {
+                    // There is a race condition here such that it's possible the method could return
+                    // 1 (or more) but no threads would be available a short time later when the scheduler
+                    // calls RunInThread. This could be avoided by 'reserving' threads for callers of
+                    // BlockForAvailableThreads, but that would complicate this code and nothing should
+                    // break functionally if threads are used for other tasks in between BlockForAvailableThreads
+                    // being called and RunInThread being called. 
+                    // 
+                    // The window of opportunity for such a race should be very small (unless the scheduler takes
+                    // a very long time to call RunInThread).
+                    //
+                    // In the worst case, RunInThread will just wait
+                    // for the next thread and clustered scenarios may experience some imbalanced loads.
+                    concurrencySemaphore.Wait(shutdownCancellation.Token);
+                    return 1 + concurrencySemaphore.Release();
+                }
+                catch (OperationCanceledException)
+                {
+                }
+            }
+
+            return 0;
+        }
+
+        /// <summary>
+        /// Schedules an action to run (using the task scheduler) as soon as concurrency rules allow it
+        /// </summary>
+        /// <param name="runnable">The action to be executed</param>
+        /// <returns>True if the task was successfully scheduled, false otherwise</returns>
+        public bool RunInThread(Action runnable)
+        {
+            if (runnable == null || !isInitialized || shutdownCancellation.IsCancellationRequested) return false;
+
+            // Acquire the semaphore (return false if shutdown occurs while waiting)
+            try
+            {
+                concurrencySemaphore.Wait(shutdownCancellation.Token);
```

Comment:
Does IThreadPool require that RunInThread block the current thread until a thread pool thread is available?  Or does it just require that the scheduler limit the actual concurrency?  If the latter, I'd do this wait asynchronously, to avoid blocking the current thread...